rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -0,0 +1,305 @@
+        }
+
+        def storeWaitingContinuation(replays: ReplayData, consume: Consume): None.type = {
+          store.putWaitingContinuation(
+            txn,
+            channels,
+            WaitingContinuation(patterns, continuation, persist, consume))
+          for (channel <- channels) store.addJoin(txn, channel, channels)
+          logger.debug(s"""|consume: no data found,
+                           |storing <(patterns, continuation): ($patterns, $continuation)>
+                           |at <channels: $channels>""".stripMargin.replace('\n', ' '))
+          replayData.put(replays)
+          None
+        }
+
+        def handleMatches(mats: Seq[DataCandidate[C, A]],
+                          replays: ReplayData,
+                          consumeRef: Consume,
+                          comms: Multiset[COMM]): Option[(K, Seq[A])] = {
+          store.eventsCounter.registerConsumeCommEvent()
+          val commRef = COMM(consumeRef, mats.map(_.datum.source))
+          assert(comms.contains(commRef), "COMM Event was not contained in the trace")
+          mats
+            .sortBy(_.datumIndex)(Ordering[Int].reverse)
+            .foreach {
+              case DataCandidate(candidateChannel, Datum(_, persistData, prodRef), dataIndex) =>
+                if (!persistData) {
+                  store.removeDatum(txn, Seq(candidateChannel), dataIndex)
+                }
+                replays.removeBinding(prodRef, commRef)
+            }
+          logger.debug(s"consume: data found for <patterns: $patterns> at <channels: $channels>")
+          replayData.put(replays.removeBinding(consumeRef, commRef))
+          Some((continuation, mats.map(_.datum.a)))
+        }
+
+        logger.debug(s"""|consume: searching for data matching <patterns: $patterns>
+                         |at <channels: $channels>""".stripMargin.replace('\n', ' '))
+
+        val consumeRef = Consume.create(channels, patterns, continuation, persist)
+        val replays    = replayData.take()
+
+        replays.get(consumeRef) match {
+          case None =>
+            runMatcher(None) match {
+              case None =>
+                storeWaitingContinuation(replays, consumeRef)
+              case Some(_) =>
+                val msg = "untraced event resulted in a comm event"
+                logger.error(msg)
+                throw new ReplayException(msg)
+            }
+          case Some(comms) =>
+            val maybeDataCandidates: Option[Seq[DataCandidate[C, A]]] =
+              comms
+                .iterator()
+                .asScala
+                .map { comm: COMM =>
+                  runMatcher(Some(comm))
+                }
+                .collectFirst {
+                  case Some(x) => x
+                }
+
+            maybeDataCandidates match {
+              case None =>
+                storeWaitingContinuation(replays, consumeRef)
```

Comment:
Btw do you not need to erase the element from the replay table?

Reply:
Edit: This might not be necessary as even if your replay table isn't empty at the end, the root hash will probably be the same...but I got to think a bit.

Like if you have a comm with

`x!(5) | for (_ <- x) { Nil }`

this goes in the table as

```
[[ x!(5) ]] => [CommEvent(x!(5) , for (_ <- x) { Nil }]
[[ for (_ <- x) { Nil } ]] => [CommEvent(x!(5) , for (_ <- x) { Nil }]
```

right?

So even if you're still waiting for the other pair of the comm, you need to decrement that entry from the MultiSet in case you have multiple. This way your replay table in the end would be completely empty.

Reply:
Thanks for pointing this out.  I will update.